File: LeetCode_每日一题/2023-05/20_1373. 二叉搜索子树的最大键值和.py
'''
Author: hakusai
Date: 2023-05-20 20:38:57
LastEditTime: 2023-05-20 20:40:22
'''
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache, reduce, cmp_to_key
from itertools import accumulate, combinations, permutations, product
from heapq import nsmallest, nlargest, heapify, heappop, heappush
from bisect import bisect_left, bisect_right
from math import factorial, gcd
from cmath import inf
from typing import List

class Solution:
    def maxSumBST(self, root: Optional[TreeNode]) -> int:
        ans = 0  # 二叉搜索树可以为空

        def dfs(node: Optional[TreeNode]) -> Tuple:
            if node is None:
                return inf, -inf, 0
            l_min, l_max, l_sum = dfs(node.left)  # 递归左子树
            r_min, r_max, r_sum = dfs(node.right)  # 递归右子树
            x = node.val
            if x <= l_max or x >= r_min:  # 不是二叉搜索树
                return -inf, inf, 0
            s = l_sum + r_sum + x  # 这棵子树的所有节点值之和
            nonlocal ans
            ans = max(ans, s)
            return min(l_min, x), max(r_max, x), s
        dfs(root)
        return ans

File: LeetCode_每日一题/2023-05/20_1373. 二叉搜索子树的最大键值和.java
/*
 * @Author: hakusai
 * @Date: 2023-05-20 20:39:15
 * @LastEditTime: 2023-05-20 22:06:05
 * @Description: 
 */

import java.util.*;
import java.io.*;
import java.math.*;

class Solution {
    private int ans;

    public int maxSumBST(TreeNode root) {
        dfs(root);
        return ans;
    }

    private int[] dfs(TreeNode node) {
        if (node == null)
            return new int[] { Integer.MAX_VALUE, Integer.MIN_VALUE, 0 };
        int[] left = dfs(node.left); // 递归左子树
        int[] right = dfs(node.right); // 递归右子树
        int x = node.val;
        if (x <= left[1] || x >= right[0]) // 不是二叉搜索树
            return new int[] { Integer.MIN_VALUE, Integer.MAX_VALUE, 0 };
        int s = left[2] + right[2] + x; // 这棵子树的所有节点值之和
        ans = Math.max(ans, s);
        return new int[] { Math.min(left[0], x), Math.max(right[1], x), s };
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
    }
}
